File: 307-range-sum-query-mutable/range-sum-query-mutable.py
class Fenwick:
    def __init__(self, nums):
        self.n = len(nums)
        self.nums = [0] + nums
        self.tree = [0] * (self.n + 1)
        for i in range(1, self.n + 1):
            self._insert(i, self.nums[i])

    def _insert(self, idx, num):
        while idx <= self.n:
            self.tree[idx] += num
            idx += idx & -idx

    def update(self, idx, num):
        idx += 1
        diff = num - self.nums[idx]
        self.nums[idx] = num
        self._insert(idx, diff)

    def _range(self, idx):
        idx += 1
        res = 0
        while idx > 0:
            res += self.tree[idx]
            idx -= idx & -idx
        return res
        
    def range_sum(self, l, r):
        return self._range(r) - self._range(l - 1)


class NumArray:

    def __init__(self, nums: List[int]):
        self.fenwick = Fenwick(nums)

    def update(self, index: int, val: int) -> None:
        self.fenwick.update(index, val)

    def sumRange(self, left: int, right: int) -> int:
        return self.fenwick.range_sum(left, right)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: 307-range-sum-query-mutable/range-sum-query-mutable.cpp
typedef vector<int> vi;

class SegmentTree {
    private:
        vi nums, sg;
        int n;
        int left(int x) {return (x << 1) + 1;}
        int right(int x) {return (x << 1) + 2;}
        void build(int p, int l, int r){
            if (l == r) {
                sg[p] = nums[l];
                return;
            }
            int mid = l + (r - l) / 2;
            int le = left(p), ri = right(p);
            build(le, l, mid); build(ri, mid + 1, r);
            sg[p] = sg[le] + sg[ri];
        }
        int queryHelper(int p, int l, int r, int L, int R) {
            if (l > R || r < L) return 0;
            if (L <= l && r <= R) return sg[p];
            int le = left(p), ri = right(p);
            int mid = l + (r - l) / 2;
            int s1 = queryHelper(le, l, mid, L, R);
            int s2 = queryHelper(ri, mid + 1, r, L, R);
            return s1 + s2;            
        }
        void updateHelper(int p, int idx, int val, int l, int r) {
            if (l == r) {
                sg[p] = val;
                return;
            }
            int mid = l + (r - l) / 2;
            if (idx <= mid) updateHelper(left(p), idx, val, l, mid);
            else updateHelper(right(p), idx, val, mid + 1, r);
            sg[p] = sg[left(p)] + sg[right(p)];
        }
    public:
        SegmentTree(vi nums) {
            this->nums = nums;
            n = nums.size();
            sg.assign(4 *n, 0);
            build(0, 0, n - 1);
        }

        int query(int left, int right) {
            return queryHelper(0, 0, n - 1, left, right);
        }

        void update(int index, int val) {
            updateHelper(0, index, val, 0, n - 1);
            nums[index] = val;
        }

};

class NumArray {
public:
    SegmentTree *sg;
    NumArray(vector<int>& nums) {
        sg = new SegmentTree(nums);
    }
    
    void update(int index, int val) {
        sg->update(index, val);
    }
    
    int sumRange(int left, int right) {
        return sg->query(left, right);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */